noBrokenSymlinks: check for unreadable symlinks

Rhys-T opened the pull request:
Resolves #380681.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Rhys-T committed:
noBrokenSymlinks: check for unreadable symlinks

Resolves #380681.

Rhys-T commented:
I haven't tested this locally yet because it wants to Rebuild The Everything (well, it _is_ a stdenv change) and I don't have that much free space.

wolfgangwalther reviewed (dismissed):
Please add a test for this in `pkgs/test/stdenv/no-broken-symlinks.nix` and target `staging`.

Rhys-T committed:
noBrokenSymlinks: add test for unreadable symlinks; fix existing tests

Rhys-T commented:
Okay, I _think_ I did that right this time…? Do you want me to un-draft-ify it?

ConnorBaker reviewed (approved):
I haven’t run the tests, but content-wise this gets my approval! Thank you for the improvement :)

Rhys-T commented:
Is it a problem that I'm rebuilding All The Things even on Linux, where the thing I'm testing for can't even happen? If so, I might be able to split it into a separate hook that only runs on Darwin and other systems that support symlink permissions.

ConnorBaker commented:
I think the setup hook should be cross-platform, and I’m not aware of a way you’d be able to avoid making a copy of it for the non-Linux case, so I’m happy to see the change here.
Rebuilding everything is the cost of updating setup hooks used by `stdenv` — in cases where the change doesn’t impact the contents of the output (like for Linux), content-addressed derivations should be able to prevent further rebuilds, but we’re not there yet :)

Rhys-T commented:
Got it. Just wanted to check - it felt weird to be sending a PR that caused that many rebuilds without even having tested it, especially when a lot of those rebuilds weren't actually accomplishing anything.

> I’m not aware of a way you’d be able to avoid making a copy of it for the non-Linux case

The closest thing I can come up with would be to leave `no-broken-symlinks.sh` only counting dangling/reflexive links, and add a similar `no-unreadable-symlinks.sh` hook to `stdenv` on only the affected platforms… but wait: that _still_ requires rebuilding on Linux, because `no-broken-symlinks.sh` still has to be fixed to make it skip any links it can't read instead of dying. So never mind, I guess.

Rhys-T commented:
It's showing that it's skipping some checks, and it kind of looks like it's interpreting the `noBrokenSymlinks: ` prefix on the PR title (or possibly the commit message) as a package name and failing to find it. Should I have named it differently? On the original issue, I went with `stdenv: ` since that was the closest thing I could find, but I forgot to do that here.

Rhys-T commented:
Hold on, I managed to set up to test it, and something's going wrong. I can confirm that GNU Coreutils `chmod` has the `-h` option even on Linux, but for some reason in testing [I'm getting an error saying that option doesn't exist](https://github.com/Rhys-T/nur-packages/actions/runs/13475419596/job/37653991541#step:8:60896). I think it's bootstrapping using Busybox or something - it's getting an error from `/nix/store/kaajn8nlqn0m370k5km7pikr5ky7ji08-bootstrap-stage2-stdenv-linux/setup`. I'll have to figure out how to handle that case.

Rhys-T committed:
noBrokenSymlinks: use umask instead of chmod to make unreadable symlinks

The stage2 stdenv uses Coreutils 9.4, which doesn't have `chmod -h` yet.

reckenrode commented:
> No, it's Coreutils, it's just an older version - it's bootstrapping using 9.4, and `-h` was only added in 9.5. Meanwhile, [the macOS test is failing somewhere else entirely](https://github.com/Rhys-T/nur-packages/actions/runs/13475419596/job/37653991672#step:8:108974), in code that shouldn't have _anything_ to do with these changes - it's claiming `ffi_prep_closure` is undeclared while it's trying to build Python's `ctypes` module!

Fixed by https://github.com/NixOS/nixpkgs/pull/383376.

Rhys-T committed:
noBrokenSymlinks: fix tests on Linux

`mkUnreadableSymlink` was exiting before the hook could print its
messages, meaning that the combined 'broken' test wasn't getting the
symlink counts.

Rhys-T commented:
[The Linux test is fixed.](https://github.com/Rhys-T/nur-packages/actions/runs/13478276250/job/37660193887) [The macOS one is still running…](https://github.com/Rhys-T/nur-packages/actions/runs/13478276250/job/37662972781)

Rhys-T commented:
Okay, slight complication. There are several other hooks that do their own `readlink`s before `no-broken-symlinks` gets called, and those fail silently too, meaning the updated `no-broken-symlinks` never gets a chance to say anything (and its tests fail as a result).

I could fix up each individual one to skip such links, but I'm wondering: Would it be better to split the unreadable-symlink check into its own hook that runs first? That would protect any new hooks that get added later from unreadable symlinks too, instead of them causing silent failures in a different hook's test like they are right now.

If I do go with that approach, would you like me to do it here, or in a new PR?

ConnorBaker commented:
Oof.

Here are three opinions I hold:

- hooks should be idempotent
- broken symlinks (unreadable, dangling, or reflexive) are a correctness issue
- hooks should not introduce correctness issues

I don’t think it makes sense to split the unreadable symlink check into its own hook. Instead, I think we should run the broken symlink hook before the others.

Thoughts?

As a greater issue which I don’t think we need to address now: moving the hook earlier means it doesn’t help if a hook which runs later introduces broken symlinks. As a fix for that, if hooks are idempotent, you can run them repeatedly. But this is a problem for later.

Rhys-T commented:
I'm fairly new to working on the core of Nixpkgs like this, but that all sounds reasonable to me. Does it need to be renamed to be alphabetically first, or just moved to the top of `defaultNativeBuildInputs` and `pkgs/test/stdenv/hooks.nix`?

ConnorBaker commented:
Should just need to be moved earlier in the list.

Rhys-T committed:
noBrokenSymlinks: move hook to beginning of `defaultNativeBuildInputs`

This protects the rest of the standard setup hooks from unreadable and
other broken symlinks. (Otherwise, they can choke on the `readlink` step
and fail silently before `no-broken-symlinks.sh`` is even reached.)

Rhys-T commented:
Okay, [testing it now](https://github.com/Rhys-T/nur-packages/actions/runs/13512197666).

Rhys-T commented:
Okay, I've _almost_ got it. The problem now is the `pass-*-allowed` versions of the test, on macOS/Darwin. The `dontCheckForBrokenSymlinks` argument successfully skips the `noBrokenSymlinks` hook… which allows the other hooks to find the unreadable symlink and die during `readlink`, causing the test to fail. This applies to the `unreadable-symlink` test and the combined `broken-symlinks` test, in both the `absolute` and `relative` variants. Any suggestions?

ConnorBaker commented on pkgs/test/stdenv/hooks.nix:
```diff
@@ -7,6 +7,7 @@
 # ordering should match defaultNativeBuildInputs
 
 {
+  no-broken-symlinks = import ./no-broken-symlinks.nix { inherit stdenv lib pkgs; };
```
For the sake of a smaller diff, you can leave this line where it was previously -- keys in attribute sets are sorted lexicographically, so it doesn't matter where you move this to.

Rhys-T replied:
Okay - I saw `# ordering should match defaultNativeBuildInputs` a few lines above it, so I wasn't sure.

Should I amend that commit to remove that change, or add yet another commit?

ConnorBaker replied:
Ah... if that's the case, then I suppose it's correct that you moved it! Disregard my comment :)

ConnorBaker commented on pkgs/test/stdenv/no-broken-symlinks.nix:
```diff
@@ -17,6 +17,25 @@ let
+      umask 777
+      ln -s${optionalString (!absolute) "r"} "$out/unreadable-symlink-target" "$out/unreadable-symlink"
+    )
+    if readlink "$out/unreadable-symlink" >/dev/null 2>&1; then
+      nixErrorLog "symlink permissions not supported"
+      ${optionalString failIfUnsupported
+        # Postpone the failure until after no-broken-symlinks.sh has a chance to print its messages
+        "postFixupHooks+=('exit 1')"
+      }
+    fi
```
Don't try to read the symlink -- the function should just produce one which is unreadable, not try to read it.

Rhys-T replied:
The idea there was to check if I had managed to make it unreadable in the first place - if not, I'm on a platform like Linux that doesn't support permissions for symlinks.

ConnorBaker replied:
`stdenv.hostPlatform` provides booleans like `isLinux` and `isDarwin`; I would condition including the check using those rather than having the above in the test.

ConnorBaker reviewed (changes requested):
Should be good to go after requested changes are addressed.

ConnorBaker commented:
I think the function producing the unreadable symlink needs to be reduce so it does exactly what the name says -- it should only produce the symlink, not try to read it.

Broken symlinks are a correctness issue -- it's expected that trying to interact with these broken symlinks should fail, so the tests shouldn't interact with them. The tests only exist to catch and warn about their presence.

Disabling the hook is a way of communicating "I know there are correctness issues with the output(s) but I don't care" so I think the behavior of failing when trying to actually interact with the broken symlinks is valid and that we just need to avoid doing so in the tests.

Rhys-T commented:
> I think the function producing the unreadable symlink needs to be reduce so it does exactly what the name says -- it should only produce the symlink, not try to read it.

The function may not be the best place for that, but something in the test needs to see whether it's on a platform that supports symlink permissions in the first place - otherwise the Linux version of the test will "fail to fail properly" because `noBrokenSymlinks` won't have anything to report there (symlinks are always readable).

> Disabling the hook is a way of communicating "I know there are correctness issues with the output(s) but I don't care" so I think the behavior of failing when trying to actually interact with the broken symlinks is valid and that we just need to avoid doing so in the tests.

Trying to make sure I'm understanding correctly: Should I remove the `pass-*-allowed` test for unreadable symlinks (and the corresponding part of the combined broken-symlinks test) because it can't actually succeed, or should it be changed to expect an error, just a _different_ error than the normal `fail` version? Or is there some other option that I'm missing?

ConnorBaker commented:
> The function may not be the best place for that, but something in the test needs to see whether it's on a platform that supports symlink permissions in the first place - otherwise the Linux version of the test will "fail to fail properly" because `noBrokenSymlinks` won't have anything to report there (symlinks are always readable).

See my comment on `stdenv.hostPlatform.isLinux`, which you can use to conditionally add the test for unreadable symlinks.

> Trying to make sure I'm understanding correctly: Should I remove the `pass-*-allowed` test for unreadable symlinks (and the corresponding part of the combined broken-symlinks test) because it can't actually succeed, or should it be changed to expect an error, just a different error than the normal fail version? Or is there some other option that I'm missing?

The `pass-*-allowed` tests should succeed because the tests should only verify behavior of the hook: in the case where the hook is disabled and broken symlinks exist, we should expect success (because the hook does not run and so the build does not fail).

Rhys-T commented:
> See my comment on `stdenv.hostPlatform.isLinux`, which you can use to conditionally add the test for unreadable symlinks.

Got it. So far it sounds like Darwin and the *BSDs are the ones where I need to worry about symlink permissions, so I'll set it up to run on those for now.

> The `pass-*-allowed` tests should succeed because the tests should only verify behavior of the hook: in the case where the hook is disabled and broken symlinks exist, we should expect success (because the hook does not run and so the build does not fail).

Well, that's the thing: the `pass-*-allowed` tests _should_ succeed, but they aren't. The test only disables the `noBrokenSymlinks` hook, not any of the others - so those other hooks end up getting `readlink` errors.

ConnorBaker commented:
Ah, in that case I’d recommends disabling or marking them broken on non-Linux systems.

Rhys-T commented:
Is there any point in even having the `pass` one on Linux? There can never be unreadable symlinks to ignore there, so there's not much for it to test. I guess it at least confirms that the hook isn't breaking things just by being there, though…

Also, for the combined `broken-symlinks` tests: The current versions in Nixpkgs test the case of having both dangling and reflexive symlinks in the same build, and work on both platforms. I had added a third, unreadable link into that test case. What should I do about the `pass` versions on non-Linux systems? Should I mark them broken like the unreadable-only ones? Simplify them back to just dangling+reflexive on those platforms, so it can at least do as much testing as it does now? Leave them as just dangling+reflexive everywhere and not worry about unreadables for the combo test?

Sorry for all the questions - I just really want to make sure I'm doing this right, since I'm changing a fairly major component of Nixpkgs.

ConnorBaker commented:
No worries! Happy to try to answer them :)

To correct something I wrote earlier -- instead of marking a test as broken, we should instead be populating `meta.badPlatforms`. So for instance, if you have a negative test case where you make an unreadable symlink and expect it to fail, you would set `meta.badPlatforms = platforms.linux`.

I'd say keep the `pass` test for Linux just to make sure the hook isn't breaking in some scary way.

For the combined tests, rather than adding the unreadable symlink to them, I'd recommend making a new combined test which includes the unreadable symlink and setting `meta.badPlatforms = platforms.linux` -- that way, the current combined tests stay unchanged and non-linux systems have a new test.

Does that help?

Rhys-T commented:
I think so, thanks. So to summarize what I'm getting from this:

| Symlink type                         | Test type      | Platform     | Action                                                                                          |
| ------------------------------------ | -------------- | ------------ | ----------------------------------------------------------------------------------------------- |
| `dangling`, `reflexive`              | `fail`, `pass` | any          | **Run** - leave these doing what they do now                                                    |
| `unreadable`                         | `fail`         | Linux        | **Skip** (`badPlatforms`) since I can't make a symlink fail that way there                      |
| `unreadable`                         | `pass`         | Linux        | **Run** to make sure the hook isn't totally broken (it won't have anything to report, though)   |
| `unreadable`                         | `fail`         | Darwin/\*BSD | **Run** to make sure we can catch unreadable symlinks                                           |
| `unreadable`                         | `pass`         | Darwin/\*BSD | **Skip** (`badPlatforms`) because even though this hook will be okay, later ones will error out |
| `broken`                             | `fail`, `pass` | any          | **Run** - but leave it only testing `dangling` and `reflexive`, like it does now                |
| **New**: `all-broken` (or something) | `fail`, `pass` | any          | **Run/Skip in the same cases as `unreadable`**, and test for all three types of broken symlinks |

Does that sound right?

Rhys-T committed:
noBrokenSymlinks: set `meta.badPlatforms` for tests

Some platforms implement permissions for symlinks, while others -
including Linux - ignore them. As a result, testing this hook's handling
of unreadable symlinks requires careful attention to which kind of
platform we're on. See the comments by `meta.badPlatforms` for details.

Rhys-T commented:
Does this look right? I couldn't find any other example of using `meta.badPlatforms` on tests. I did see some `meta.broken`s in `pkgs/test/cc-wrapper/hardening.nix`, so as long as everything is using `lib.meta.availableOn` or something else like that, I guess `meta.badPlatforms` should work too. Not entirely clear on how to test it though. It looks like the `cc-wrapper` tests have a `supported` test derivation that pulls in just the non-broken tests in that group as dependencies. Do I need to add something similar for `no-broken-symlinks`?

Rhys-T commented:
(Oops… sorry, I didn't realize it was going to request another batch of reviewers each time I marked it ready.)

ConnorBaker commented:
I realize that setting `meta.badPlatforms` causes evaluation errors, so one can't do something like

```console
nix-build -A tests.stdenv.hooks.no-broken-symlinks
```

to run all the tests.

That shouldn't be a problem for Hydra due to the way it performs evaluation, but it is inconvenient if you're not using `nix-eval-jobs` to get the derivations to realize.

I don't know what's recommended with respect to attribute set consistency across platforms -- I've previously been told that packages should be present across platforms even if they are unsupported because missing attributes issue an `abort` instead of a `throw` (which can be caught).

That said, I'm happy with this as-is.

Anyone have any objections to merging?

ConnorBaker commented:
In that case, @Rhys-T would you mind either implementing something akin to the `supported` test derivation you mentioned earlier, or using something like `optionalAttrs` to conditionally add the tests to the attribute set depending on `lib.meta.availableOn`?

Rhys-T commented:
Sorry. I started working on this the other day, but then Life Happened and I never got back to it.

So far, what I have locally is a new `tests.stdenv.hooks.no-broken-symlinks.supported` attrset, containing just the subset of those tests that aren't broken on the current system. Will that work okay, or does it need to be a full derivation like [`tests.cc-wrapper.supported`](https://github.com/NixOS/nixpkgs/blob/6e15303353a1a3977447c6c26a9b9c8cadab13d5/pkgs/test/default.nix#L62-L73)?

One thing I'm still not clear on from the documentation (which might just mean I don't know what docs I should be looking at) is whether there's any particular structure to `tests` that any of the tooling is expecting. I'm currently leaving all the individual tests available under `tests.stdenv.hooks.no-broken-symlinks` in an attempt to maintain backwards compatibility, but could that actually cause problems instead since not all those tests are runnable? Also, while `tests.stdenv.hooks` is marked with `recurseIntoAttrs`, `[…].no-broken-symlinks` isn't. Should the `supported` attrset be flagged that way? Everything else under `[…].hooks` is a single derivation, so I don't really have anything to compare against.

ConnorBaker commented:
Excellent questions!

> So far, what I have locally is a new `tests.stdenv.hooks.no-broken-symlinks.supported` attrset, containing just the subset of those tests that aren't broken on the current system. Will that work okay, or does it need to be a full derivation like [`tests.cc-wrapper.supported`](https://github.com/NixOS/nixpkgs/blob/6e15303353a1a3977447c6c26a9b9c8cadab13d5/pkgs/test/default.nix#L62-L73)?

Having an attribute set is fine -- it doesn't need to be a full derivation. I think people running tests are either using some combination of `nix-eval-jobs` and `nix build` to handle building the entirety of an attribute set, or `nix-build -A` because it already handles that.

> One thing I'm still not clear on from the documentation (which might just mean I don't know what docs I should be looking at) is whether there's any particular structure to `tests` that any of the tooling is expecting.

As far as I'm aware, there's no particular structure expected. That said, I've not done a survey of tools to figure out if any of them make any assumptions, and I'm not aware of any tools specific to `tests`.

> I'm currently leaving all the individual tests available under `tests.stdenv.hooks.no-broken-symlinks` in an attempt to maintain backwards compatibility, but could that actually cause problems instead since not all those tests are runnable?

Yep! If someone were to want to build the tests for `stdenv`'s setup hooks, they might try to run

```bash
nix-build -A tests.stdenv.hooks
```

Unfortunately, I think the current approach would cause that to fail since some of the attributes under `tests.stdenv.hooks.no-broken-symlinks` won't build.

Feel free to re-organize the test file however makes most sense -- I'd suggest doing something like

```nix
{stdenv, etc, ...}: # callPackage arguments
let
  inherit (lib.attrsets) optionalAttrs recurseIntoAttrs;
  inherit (stdenv.hostPlatform) isLinux isDarwin;
  # test utility functions, etc.
in
# Common tests
recurseIntoAttrs {
  blah = mkTestWhichWhichWorksOnEverything {};
}
# Linux-specific tests
// optionalAttrs isLinux {
  linuxFoo = mkTestWhichWorksOnLinux {};
}
# Darwin-specific tests
// optionalAttrs isDarwin {
  darwinBar = mkTestWhichWorksOnDarwin {};
}
```

> Also, while `tests.stdenv.hooks` is marked with `recurseIntoAttrs`, `[…].no-broken-symlinks` isn't. Should the `supported` attrset be flagged that way? Everything else under `[…].hooks` is a single derivation, so I don't really have anything to compare against.

Ah, `recurseIntoAttrs`...

As I understand it, every attribute set which contains derivations (including transitively) should have `recurseIntoAttrs` applied to it, so yes, `no-broken-symlinks` should have `recurseIntoAttrs` applied to it. It's used in a few places, like Nix[^1], `nix-eval-jobs`[^2], and even Nixpkgs[^3], mostly to figure out whether to continue recursing (setting `recurseForDerivations = false` is a handy way to prevent infinitely recursing into a self-referential child attribute).

Likewise, if you were to make an attribute set nested within `no-broken-symlinks` which contained derivations, it too should have `recurseIntoAttrs` applied to it.

Please let me know if that helps or you have any more questions!

[^1]: https://github.com/NixOS/nix/blob/2c1447b7f2ed523631471b9a232b462d5e7fb075/src/libexpr/get-drvs.cc#L388

[^2]: https://github.com/nix-community/nix-eval-jobs/blob/4b392b284877d203ae262e16af269f702df036bc/src/worker.cc#L262

[^3]: https://github.com/NixOS/nixpkgs/blob/d325772e9e7213ee7e003dba641e664899c7b790/pkgs/top-level/release-lib.nix#L169-L177

Rhys-T committed:
noBrokenSymlinks: use `lib.optionalAttrs` instead of `meta.badPlatforms`

Having unbuildable tests in the attrset makes it difficult to ask Nix to
build all the working ones at once. See discussion in #380683. So
instead of flagging them, let's just remove them entirely.

Rhys-T committed:
noBrokenSymlinks: apply `lib.recurseIntoAttrs` to set of tests

For consistency with the outer `tests.stdenv.hooks` attrset.

Rhys-T commented:
[Working on running the tests for this latest version now](https://github.com/Rhys-T/nur-packages/actions/runs/13686124269/job/38269823437), but I _think_ I've finally got it ready at this point…?

ConnorBaker commented:
Wonderful! I was able to verify it works on `x86_64-linux`, `aarch64-darwin`, and `x86_64-darwin`.

Thank you so much for taking this on!
